boa_examples/src/bin/modules.rs

```diff
@@ -0,0 +1,120 @@
+use std::{error::Error, path::Path};
+
+use boa_engine::{
+    builtins::promise::PromiseState,
+    module::{ModuleLoader, SimpleModuleLoader},
+    object::FunctionObjectBuilder,
+    Context, JsError, JsNativeError, JsValue, Module, NativeFunction,
+};
+use boa_parser::Source;
+
+// This example demonstrates how to use Boa's module API
+fn main() -> Result<(), Box<dyn Error>> {
+    // A simple module that we want to compile from Rust code.
+    const MODULE_SRC: &str = r#"
+        import { pyth } from "./trig.mjs";
+        import * as ops from "./operations.mjs";
+
+        export let result = pyth(3, 4);
+        export function mix(a, b) {
+            return ops.sum(ops.mult(a, ops.sub(b, a)), 10);
+        }
+    "#;
+
+    // This can be overriden with any custom implementation of `ModuleLoader`.
+    let loader = &SimpleModuleLoader::new("./boa_examples/scripts/modules")?;
+    let dyn_loader: &dyn ModuleLoader = loader;
+
+    // Just need to cast to a `ModuleLoader` before passing it to the builder.
+    let context = &mut Context::builder().module_loader(dyn_loader).build()?;
+    let source = Source::from_reader(MODULE_SRC.as_bytes(), Some(Path::new("./main.mjs")));
+
+    // Can also pass a `Some(realm)` if you need to execute the module in another realm.
+    let module = Module::parse(source, None, context)?;
+
+    // Don't forget to insert the parsed module into the loader itself! Since the root module
+    // is not automatically inserted by the `ModuleLoader::load_imported_module` impl.
+    //
+    // Simulate as if the "fake" module is located in the modules root, just to ensure that
+    // the loader won't double load in case someone tries to import "./main.mjs".
+    loader.insert(
+        Path::new("./boa_examples/scripts/modules")
+            .canonicalize()?
+            .join("main.mjs"),
+        module.clone(),
+    );
+
+    // The lifecycle of the module is tracked using promises, which can be a bit cumbersome
+    // for simple uses but that use case is better suited by the `Module::load_link_evaluate` method.
+    // This does the full version for demonstration purposes.
+    //
+    // parse -> load -> link -> evaluate
```

Comment:
A bit nitpicky on the example docs comment, but could the comments/documentation on this section of the example be a bit more step-by-step of what's happening?

Something like:

```rust
    // 
    // The lifecycle goes through 4 main steps: parse -> load -> link -> evaluate
    //
    // Begin with the parsed module
    let promise_result = module
        // load the module
        .load(context)
        // Use a `NativeFunction` to links the module.
        .then(
            Some(
                FunctionObjectBuilder::new(
                    context,
                    NativeFunction::from_copy_closure_with_captures(
                        |_, _, module, context| {
                            module.link(context)?;
                            Ok(JsValue::undefined())
                        },
                        module.clone(),
                    ),
                )
                .build(),
            ),
            None,
            context,
        )?
        // Finally, use a `NativeFunction` to evaluate the module.
        .then(
            Some(
                FunctionObjectBuilder::new(
                    context,
                    NativeFunction::from_copy_closure_with_captures(
                        |_, _, module, context| Ok(module.evaluate(context).into()),
                        module.clone(),
                    ),
                )
                .build(),
            ),
            None,
            context,
        )?;
```